test/e2e/karmadactl_test.go: test register command

mohamedawnallah opened the pull request:
**What type of PR is this?**

/kind failing-test


**Which issue(s) this PR fixes**:

Part of #4544 

**Context**

Dependent on #5429

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required.
-->
```release-note
NONE
```

mohamedawnallah commented:
Currently this test case `should register a cluster with CA verification` fails with the following error:
```
  [FAILED] Unexpected error:
      <exec.CodeExitError>:
      error running /home/karmada/go/bin/karmadactl --kubeconfig=/home/karmada/.kube/member-e2e-gfj.config register 172.18.0.8:6443 --token bxfkm7.0xx94rzy1mda3iun --discovery-token-ca-cert-hash sha256:49514bb5d379ef5940a6e22709bfba9085950a0e54d2811c642fb75e4bdfd76c --cluster-name member-e2e-gfj --karmada-agent-image docker.io/karmada/karmada-agent:latest --ca-cert-path /tmp/karmada-e2e-301700272/ca.crt:
      Command stdout:
      [preflight] Running pre-flight checks
      [preflight] All pre-flight checks were passed
      [karmada-agent-start] Waiting to perform the TLS Bootstrap
      [karmada-agent-start] Waiting to construct karmada-agent kubeconfig

      stderr:
      Unable to connect to the server: dial tcp: lookup member-e2e-gfj-control-plane on 127.0.0.53:53: server misbehaving

      error:
      exit status 1
      {
          Err: <*errors.errorString | 0xc000062060>{
              s: "error running /home/karmada/go/bin/karmadactl --kubeconfig=/home/karmada/.kube/member-e2e-gfj.config register 172.18.0.8:6443 --token bxfkm7.0xx94rzy1mda3iun --discovery-token-ca-cert-hash sha256:49514bb5d379ef5940a6e22709bfba9085950a0e54d2811c642fb75e4bdfd76c --cluster-name member-e2e-gfj --karmada-agent-image docker.io/karmada/karmada-agent:latest --ca-cert-path /tmp/karmada-e2e-301700272/ca.crt:\nCommand stdout:\n[preflight] Running pre-flight checks\n[preflight] All pre-flight checks were passed\n[karmada-agent-start] Waiting to perform the TLS Bootstrap\n[karmada-agent-start] Waiting to construct karmada-agent kubeconfig\n\nstderr:\nUnable to connect to the server: dial tcp: lookup member-e2e-gfj-control-plane on 127.0.0.53:53: server misbehaving\n\nerror:\nexit status 1",
          },
          Code: 1,
      }
  occurred
```

Not sure if this is dependent on this issue #4262 ?

XiShanYongYe-Chang commented:
Thanks @mohamedawnallah 
The E2E seems to have failed for all versions.
/assign @zhzhuang-zju

mohamedawnallah commented:
I've checked the logs, and there is an issue with creating the [KarmadaDir](https://github.com/karmada-io/karmada/blob/37488bd1093a2c5cf808b6cf87d601caab07a05f/pkg/karmadactl/register/register.go#L80) at `/etc/karmada` because this directory is owned by `root` and requires `sudo` privileges. I've bypassed this locally by running:

```bash
sudo -E env "PATH=$PATH" ginkgo -focus "Karmadactl register testing"
```

This allows the test cases to bypass the issue, but I think there are a couple of problems with this approach. Currently, running Ginkgo with `sudo` in the CI environment could lead to issues later on?

A potential improvement could be to add a flag, something like `--karmada-directory`, to the `karmadactl register` command, similar to the `--ca-cert-path` flag. This would provide more flexibility for the end-user by allowing the use of a temporary directory during testing, thereby solving the issue out of the box. WDYT?

XiShanYongYe-Chang commented:
Hi @mohamedawnallah based on your previous E2E and UT, I think your test organization ideas are clear and your code is clean and easy to read. I look forward to your continued contributions. ^-^

zhzhuang-zju commented:
> A potential improvement could be to add a flag, something like `--karmada-directory`, to the `karmadactl register` command, similar to the `--ca-cert-path` flag. This would provide more flexibility for the end-user by allowing the use of a temporary directory during testing, thereby solving the issue out of the box. WDYT?

that make sense. Users might have the same issue under a non-root user account.

zhzhuang-zju commented:
> Hi @mohamedawnallah, I understand that the associated blockage has been resolved by #5862 . Do you have time to push this pr forward?

Thanks for reminding this, I almost forgot about this.

mohamedawnallah commented:
> Hi @mohamedawnallah, I understand that the associated blockage has been resolved by https://github.com/karmada-io/karmada/pull/5862 . Do you have time to push this pr forward?

Thanks, @XiShanYongYe-Chang, for bringing this to my attention. I've updated the test cases following the resolution of PR #5862. The remaining issue for this PR seems to be the error: `Unable to connect to the server: dial tcp: lookup member-e2e-jnpl9-control-plane on 127.0.0.53:53: server misbehaving`, which occurs when registering a cluster using a bootstrap token. I'm not sure if this is a known issue with the `karmada register` operation or something similar to what was discussed in #4262? I would love to hear your thoughts all.

XiShanYongYe-Chang commented:
Ask @zhzhuang-zju to help take a look~

zhzhuang-zju commented:
I‘ll take a look when the ci is completed

mohamedawnallah commented:
> The remaining issue for this PR seems to be the error: Unable to connect to the server: dial tcp: lookup member-e2e-jnpl9-control-plane on 127.0.0.53:53: server misbehaving, which occurs when registering a cluster using a bootstrap token. I'm not sure if this is a known issue with the karmada register operation or something similar to what was discussed in https://github.com/karmada-io/karmada/issues/4262? I would love to hear your thoughts all.

I've solved the issue. I found that I was using the member cluster config to create the bootstrap token instead of the Karmada APIServer config file. Changing this resolved the problem.

> error validating "/tmp/deploy-4jdtb.yaml": error validating data: [apiVersion not set, kind not set]; if you choose to ignore these errors, turn validation off with --validate=false

The e2e tests for this registering operation pass locally, but the error occurred in the CI tests and seems unrelated to the changes proposed in this PR. I've rebased onto the master branch since this PR has been open for a while, but the issue still persists. Any thoughts, @zhzhuang-zju?

mohamedawnallah commented:
> The remaining issue for this PR seems to be the error: Unable to connect to the server: dial tcp: lookup member-e2e-jnpl9-control-plane on 127.0.0.53:53: server misbehaving, which occurs when registering a cluster using a bootstrap token. I'm not sure if this is a known issue with the karmada register operation or something similar to what was discussed in https://github.com/karmada-io/karmada/issues/4262? I would love to hear your thoughts all.

I've solved that issue. I found that I was using the member cluster config to create the bootstrap token instead of the Karmada APIServer config file. Changing this resolved the problem.

> error validating "/tmp/deploy-4jdtb.yaml": error validating data: [apiVersion not set, kind not set]; if you choose to ignore these errors, turn validation off with --validate=false

The e2e tests for this registering operation pass locally, but the error occurred in the CI tests and seems unrelated to the changes proposed in this PR. I've rebased onto the master branch since this PR has been open for a while, but the issue still persists. Any thoughts, @zhzhuang-zju?

zhzhuang-zju commented on test/e2e/suites/base/karmadactl_test.go:
```diff
@@ -26,6 +27,7 @@ import (
 
 	"github.com/onsi/ginkgo/v2"
 	"github.com/onsi/gomega"
+	"gopkg.in/yaml.v3"
```
@mohamedawnallah  Is it caused by the change of this YAML library, from "sigs.k8s.io/yaml" to "gopkg.in/yaml.v3"?

mohamedawnallah replied:
Great observation, @zhzhuang-zju! 👌 That's right—Visual Studio Code's Ctrl+S failed me this time! 😂

mohamedawnallah commented:
>  s: "error running /home/runner/go/bin/karmadactl --kubeconfig=/home/runner/.kube/karmada.config register 172.18.0.4:6443 --token 2mnk2l.4segshmnxwcyshc9 --discovery-token-ca-cert-hash sha256:3477db6b4299125dc026573a04ebba3ebad161593c11498a7b66e3a150685ef2 --kubeconfig=/home/runner/.kube/member-e2e-9kc55.config --cluster-name member-e2e-9kc55 --karmada-agent-image docker.io/karmada/karmada-agent:latest:\nCommand stdout:\n[preflight] Running pre-flight checks\n[preflight] All pre-flight checks were passed\n[karmada-agent-start] Waiting to perform the TLS Bootstrap\n\nstderr:\nUnable to connect to the server: dial tcp: lookup karmada-host-control-plane on 127.0.0.53:53: server misbehaving\n\nerror:\nexit status 1",

It looks like the end-to-end test cases fail due to interference with other members. Running the register test case sequentially instead of in parallel may resolve the issue.

mohamedawnallah commented:
Hey @zhzhuang-zju, How am I supposed to move this PR forward? I've already addressed the feedback and all the test cases pass :)

XiShanYongYe-Chang commented:
Good job!  Isn't it okay for us to merge it?

mohamedawnallah commented:
> Hey @zhzhuang-zju, How am I supposed to move this PR forward? I've already addressed the feedback and all the test cases pass :)

> Sorry for the delay, I'll take a look ASAP

Never mind! I understand you've been busy. Just ping me if you have any additional feedback, and I'd be happy to address it :)

zhzhuang-zju commented on test/e2e/suites/base/karmadactl_test.go:
```diff
@@ -1542,6 +1544,232 @@ func WriteYamlToFile(obj interface{}, filePath string) error {
+			output, err := cmd.ExecOrDie()
+			gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+
+			// Extract the endpoint for Karmada APIServer.
+			endpointRegex := regexp.MustCompile(`(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}):(\d{1,5})`)
+			endpointMatches := endpointRegex.FindStringSubmatch(output)
+			gomega.Expect(len(endpointMatches)).Should(gomega.BeNumerically(">=", 2))
+			karmadaAPIEndpoint = fmt.Sprintf("%s:%s", endpointMatches[1], endpointMatches[2])
+			gomega.Expect(karmadaAPIEndpoint).Should(gomega.Equal(karmadaAPIEndpointExpected))
+
+			// Extract token.
+			tokenRegex := regexp.MustCompile(`--token\s+(\S+)`)
+			tokenMatches := tokenRegex.FindStringSubmatch(output)
+			gomega.Expect(len(tokenMatches)).Should(gomega.BeNumerically(">", 1))
+			token = tokenMatches[1]
+
+			// Extract discovery token CA cert hash.
+			hashRegex := regexp.MustCompile(`--discovery-token-ca-cert-hash\s+(\S+)`)
+			hashMatches := hashRegex.FindStringSubmatch(output)
+			gomega.Expect(len(hashMatches)).Should(gomega.BeNumerically(">", 1))
+			discoveryTokenCACertHash = hashMatches[1]
```
Since the default validity period of the token is 24 hours, we can use the same token for different tests of the register command. I suggest moving this code into the beforeEach section for use by the subsequent two test suites. In this way, the function `extractAPIServerEndpoint` can also be omitted.

mohamedawnallah replied:
> Since the default validity period of the token is 24 hours, we can use the same token for different tests of the register command. I suggest moving this code into the beforeEach section for use by the subsequent two test suites

I agree with reusing the token creation logic since tokens expire after 24 hours. However, as only one test case for CA Verification uses this token, is reusing the code necessary?

> In this way, the function extractAPIServerEndpoint can also be omitted.

It looks like `extractAPIServerEndpoint` function is still needed  in `should register a cluster without CA verification` test case?

mohamedawnallah replied:
> Since the default validity period of the token is 24 hours, we can use the same token for different tests of the register command. I suggest moving this code into the beforeEach section for use by the subsequent two test suites

I agree with reusing the same token for different tests if any since tokens expire after 24 hours. However, as only one test case for CA Verification uses this token, is putting the token creation logic in `beforeEach` necessary?

> In this way, the function extractAPIServerEndpoint can also be omitted.

It looks like `extractAPIServerEndpoint` function is still needed  in `should register a cluster without CA verification` test case?

mohamedawnallah replied:
> In this way, the function extractAPIServerEndpoint can also be omitted.

Upon reviewing the testing code more closely, I now recall that I used `extractAPIServerEndpoint` to test the actual APIServerEndpoint we retrieve from `--print-register-command` against the one we already know is true inside the `kubeconfig` global variable

zhzhuang-zju commented on test/e2e/suites/base/karmadactl_test.go:
```diff
@@ -1542,6 +1544,232 @@ func WriteYamlToFile(obj interface{}, filePath string) error {
+			gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+		})
+
+		ginkgo.By(fmt.Sprintf("Set host-control-plane kubeconfig current context to %s", karmadaContext), func() {
```
We can specify the kubeconfig and context through flags, so is it necessary to specifically set the current context?

mohamedawnallah replied:
> so is it necessary to specifically set the current context?

I don't recall the exact context back then but will try it and let you know.

mohamedawnallah replied:
> We can specify the kubeconfig and context through flags,

That's right. I've removed the code for setting the context, and the register e2e tests passed. I don't recall the rationale for manually setting the context back then.

zhzhuang-zju commented:
Great job~ thanks @mohamedawnallah 
/lgtm
leave Approve to @XiShanYongYe-Chang to take a look

zhzhuang-zju commented on test/e2e/suites/base/karmadactl_test.go:
```diff
@@ -1542,6 +1542,150 @@ func WriteYamlToFile(obj interface{}, filePath string) error {
+
+			// Extract the endpoint for Karmada APIServer.
+			endpointRegex := regexp.MustCompile(`(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}):(\d{1,5})`)
+			endpointMatches := endpointRegex.FindStringSubmatch(output)
+			gomega.Expect(len(endpointMatches)).Should(gomega.BeNumerically(">=", 2))
+			karmadaAPIEndpoint = fmt.Sprintf("%s:%s", endpointMatches[1], endpointMatches[2])
```
There is an issue here. If the length of `endpointMatches` is 2, then accessing `endpointMatches[2]` will cause a panic.

We don't have the requirement to extract the IP and port separately. So, can we directly use `FindString` instead of `FindStringSubmatch`?

zhzhuang-zju replied:
@mohamedawnallah If you are going to fix this issue, you can also remove the "k8syaml" alias for the "sigs.k8s.io/yaml" package. Currently, there is no need to add this alias.

zhzhuang-zju commented:
/lgtm
cc @XiShanYongYe-Chang

XiShanYongYe-Chang reviewed (commented):
Thanks a lot
/approve
